stablehlo/compatibility/dialect/VersionNumber.h

```diff
@@ -0,0 +1,99 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef STABLEHLO_COMPATIBILITY_VERSION_NUMBER_H
+#define STABLEHLO_COMPATIBILITY_VERSION_NUMBER_H
+
+#include <string>
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Regex.h"
+#include "mlir/Support/LogicalResult.h"
+
+namespace mlir {
+namespace versionedhlo {
+
+class VersionNumber {
+  static constexpr llvm::StringLiteral CURRENT_VERSION = "0.1.0";
+  static constexpr llvm::StringLiteral MINIMUM_VERSION = "0.0.0";
+
+  public:
+    static FailureOr<VersionNumber> get(llvm::StringRef versionRef) {
+      if (failed(VersionNumber::validateVersionString(versionRef))) {
+        return failure();
+      }
+      return VersionNumber(versionRef);
+    }
+
+    static VersionNumber getCurrent() {
+      return VersionNumber(CURRENT_VERSION);
+    }
+    static VersionNumber getMinimumSupported() {
+      return VersionNumber(MINIMUM_VERSION);
+    }
+
+    /// Validate version argument is one of {current, minimum, #.#.#}
+    static LogicalResult validateVersionString(llvm::StringRef versionRef) {
+      if (versionRef == "current" || versionRef == "minimum") {
+        return success();
+      }
+
+      llvm::Regex versionRegex("^([0-9]+)\\.([0-9]+)\\.([0-9]+)$");
+      llvm::SmallVector<llvm::StringRef> matches;
+      return success(/*isSuccess=*/versionRegex.match(versionRef, &matches));
+    }
+
+    int64_t getMinorVersion() const {
+      if (version == "current")
+        return parseMinorVersion(VersionNumber::CURRENT_VERSION);
+      if (version == "minimum")
+        return parseMinorVersion(VersionNumber::MINIMUM_VERSION);
+      return parseMinorVersion(version);
+    }
+
+    bool operator<(VersionNumber const & other) {
+      return getMinorVersion() < other.getMinorVersion();
+    }
+    bool operator==(VersionNumber const& other) {
+      return getMinorVersion() == other.getMinorVersion();
+    }
+    bool operator<=(VersionNumber const & other) {
+      return *this < other || *this == other;
+    }
+
+  private:
+    VersionNumber(llvm::StringRef versionRef) : version(versionRef) {
+      assert(succeeded(VersionNumber::validateVersionString(versionRef)));
+    }
+
+    int64_t parseMinorVersion(llvm::StringRef versionRef) const {
+      // Precondition: must be x.y.z
+      auto isDot = [](char c) { return c == '.'; };
+      auto minorS = versionRef.drop_until(isDot).drop_front(1).take_until(isDot);
+      int64_t minor;
+      if (minorS.getAsInteger(/*radix=*/10, minor)) {
+        llvm_unreachable("failed to parse minor version");
+      }
+      return minor;
+    }
+
+    std::string version;
+};
```

Comment:
This cleaned things up a ton. Went with an approach where we store a `std::array<int64_t, 3>` so we can leverage array comparison operators. I.e. `std::array::operator<` uses `lexicographical_compare` by default which is exactly what we want for version numbers.

Reply:
Fixed so that all public aspects of this class are individual fields, array is an implementation detail now.

Keeping member variable  since comparisons are common and the options are either to re-implement a 3-field lexicographical compare, or construct a container to use for the std algorithm on each compare.